src/nanotron/trainer.py

```diff
@@ -432,13 +496,43 @@ def train(
                 # TODO(xrsrke): refactor using callbacks would be better
                 self.metadata.consumed_train_samples += self.global_batch_size
                 self.metadata.last_train_step = self.iteration_step
-                self.metadata.data_stages[
-                    self.metadata.last_stage_idx
-                ].consumed_train_samples += self.global_batch_size
+                self.metadata.data_stages[self.metadata.last_stage_idx].consumed_train_samples += self.global_batch_size
 
                 if (self.iteration_step - 1) % self.config.logging.iteration_step_info_interval == 0:
                     self.train_step_logs(outputs=outputs, loss_avg=loss_avg)
 
+                if (not skip_validation) and (self.iteration_step - 1) % self.val_check_interval == 0:
+                    self._update_valid_dataloader_based_on_training_stages(dataloader_valid)
+                    valid_outputs = self.validation_step(dataloader=self.current_valid_dataloader)
+
+                    if isinstance(valid_outputs[0]["loss"], torch.Tensor):
+                        loss_avg = torch.stack([_["loss"] for _ in valid_outputs]).sum()
+                        handle = dist.all_reduce(loss_avg, group=self.parallel_context.dp_pg, async_op=True,
+                                                 op=dist.ReduceOp.AVG)
+                    else:
+                        loss_avg = None
+                        handle = None
+
+                    if handle is not None:
+                        handle.wait()
+
+                    self.valid_metadata.consumed_train_samples += self.global_batch_size
+                    self.valid_metadata.last_train_step = self.iteration_step
+                    self.valid_metadata.data_stages[self.valid_metadata.last_stage_idx].consumed_train_samples += self.global_batch_size
+
+                    log_entries = [LogItem("validation_loss_avg", loss_avg, "human_format")]
+                    self.loggerwriter.add_scalars_from_list(log_entries, self.iteration_step)
```

Comment:
```suggestion
                     if dist.get_rank(self.parallel_context.world_pg) in self.logger_ranks:
                        log_entries = [LogItem("validation_loss_avg", loss_avg, "human_format")]
                        self.loggerwriter.add_scalars_from_list(log_entries, self.iteration_step)

```

When using dp>1 I got an error. Fixed it with this.